### DISCUSSION TITLE
Python's features coverage list

### DISCUSSION THREAD

**Discussion by ormeilu** [modular/modular#634] [Ideas]
## What's the problem

First of all - I really enjoy playing in Mojo right now. But only "playing" for now. 

I really want to use Mojo in production, but **there is nothing in Mojo's documentation about** _"How many of Python's features we have so far"_ and this is really annoying, because Mojo is trying to be a _"super-set"_ - so I can just Copy-Paste half of my project and it should works perfectly. 
I know that some of the features isn't made yet, team is targeted on Mojo more, than on Python and this is really good - **I respect that**

**But knowing what Mojo is able to do in _Pythonic way_ is also important.**

For example, I've tried to do the simple thing - _f-string_ (`f'hello {name}'`) and got an error. I've started to look in Issues for `f-strings` to know if it's implemented or not, and found nothing so far, only a few poor guys like me.

## How others solved that

As an average Apple M1 enjoyer and Pytorch fan, I've subscribed to [General MPS op coverage tracking issue](https://github.com/pytorch/pytorch/issues/77764) and now **I can track what can I do on my MacBook and what I can not.** **This is so helpfull when you trying to understand why this particular part doesn't work** (only because its not ready yet). You know that it doesn't work only for a short time and this feature is will be implemented soon. _**You know that developers knows about that.**_ 

## What am I asking you

It would be really great if you guys **create some roadmap of Python's features that have been implemented in Mojo** and what is not (and maybe would never be implemented?).

It can be really anything - **from GitHub issue to standalone page in Mojo's docs** (like "Coming from Python" page). I want to press `Ctrl + F` and type `f-strings` to find out that it's still under development.

Thank you, guys! You've made a great job!

**Comment by ksandvik** [comment]
Is this OK as a starting point? https://docs.modular.com/mojo/roadmap.html

**Reply by ormeilu** [reply]
Yep, looks good, but not enough to be perfect. I like the way pytorch did it - just simple list of checkboxes, that shows what is in TODO stage and what is already implemented. And items in this list is not removed after they did something - it's toggled to `DONE`. It also show a cool coverage.

And, for example - I can see in your link that comprehensions is not done yet. But, nothing about f-string, for example.

Another feature of `coverage issue` is by community some missing language features. Mojo is hard and complicated project, and developers can easily be lost. So, community's role at this can be as `de-centralized QA` that marks and tracks that for you (like another level of QA)

But yeah, that roadmap is a great `starting point`. BTW, `sharp edges` is not that obvious name for naming a `not ready yet` section, imo

**Comment by rarebreed** [comment]
I think this is very much needed.  A fair number of questions could be resolved by pointing people to some kind of Not Yet Implemented checklist.

Another somewhat related problem, is that if you only read the main blurb on mojo, it's easy to assume that Mojo is already fully _source_ compatible with python.  Examples from https://www.modular.com/mojo:

> Mojo combines the usability of Python with the performance of C,

No mention that mojo isn't fully python compatible...let's keep reading

> **Write everything in one language**
Write Python or scale all the way down to the metal.

Still nothing yet...

> Unlock Python performance

Also no mention of limitations...

> **Access the entire Python ecosystem**
Experience true interoperability with the Python ecosystem. Seamlessly intermix arbitrary libraries like Numpy and Matplotlib and your custom code with Mojo.

Oof, this makes it sound like mojo is already 100% compatible.  _Seamlessly intermix....your custom code_ and _true interoperability_ gives no hint that mojo is not source compatible with python.

> Mojo is still a work in progress, but it's available to try today via our Mojo SDK

This is the only briefest of hints that maybe mojo isn't 100% compatible yet, and there's still work to be done.

Unless you read the [python compatibility section in the Why Mojo](https://docs.modular.com/mojo/why-mojo.html#compatibility-with-python) link, the [python integration section](https://docs.modular.com/mojo/programming-manual.html#python-integration) of the Programming Manual and pretty much the entire [Roadmap and Rough Edges doc](https://docs.modular.com/mojo/roadmap.html), casual experimenters will be puzzled why things don't work, and may get discouraged.  Especially if they don't know how much is or isn't completed, and what the priorities are to make mojo a true superset of python

**Comment by iamtimdavis** [comment]
Agree this is a great idea and we'll work to figure out how to put something together soon. Appreciate the feedback 👍🏼

@dualvtable @jackos

**Comment by drunkwcodes** [comment]
[Can I Use](https://caniuse.com/ciu/index) is a good reference.

**Reply by g-otn** [reply]
Other references I could think of, for inspiration:
- [Bun.js](https://bun.sh/docs/runtime/nodejs-apis)
- [Deno](https://docs.deno.com/runtime/manual/node/compatibility)
- [ES6 Compat table](https://compat-table.github.io/compat-table/es6/)

**Comment by ksandvik** [comment]
It's a bigger project, but a comprehensive test suite with all the Python syntax possibilities and with a go/no-go flag that is exported to a web page is one option. Makes it easy to catch regressions during development, too.

I actually took a look at the CPython unit tests to run them as mojo tests, but many of them had class-based constructs so that needs to be in place in mojo first. But as a bonus mojo could then track the evolving CPython unit test suite in future with CPython changes.

**Reply by jackos** [reply]
Cool idea thanks @ksandvik

**Reply by drunkwcodes** [reply]
I'm reluctant for making Mojo a strict python superset due to possible performance compromise.

**Comment by ksandvik** [comment]
Fo using the CPython unit tests, one good option is to write a converter tool as there will be future CPython versions anyway. I will take a look, but as mentioned I think most of those required a class so that's one hurdle.
